File: src/game/achievements.ts
export interface AchievementDef {
  id: string;
  name: string;
  description: string;
  title?: string;
}

export const ACHIEVEMENTS: Record<string, AchievementDef> = {
  first_blood: {
    id: 'first_blood',
    name: 'First Blood',
    description: 'Slay your first creature.',
    title: 'Initiate',
  },
  goblin_hunter: {
    id: 'goblin_hunter',
    name: 'Goblin Hunter',
    description: 'Slay 10 goblins.',
    title: 'Goblin Hunter',
  },
  chief_slayer: {
    id: 'chief_slayer',
    name: 'Kingslayer',
    description: 'Defeat Goblin Chief Grak.',
    title: 'Giantkiller',
  },
  duelist: {
    id: 'duelist',
    name: 'Duelist',
    description: 'Win your first duel.',
    title: 'Duelist',
  },
  explorer: {
    id: 'explorer',
    name: 'Explorer',
    description: 'Visit the Crystal Cavern.',
    title: 'Pathfinder',
  },
  wealthy: {
    id: 'wealthy',
    name: 'Wealthy',
    description: 'Accumulate 500 gold.',
    title: 'Magnate',
  },
  veteran: {
    id: 'veteran',
    name: 'Veteran',
    description: 'Reach level 10.',
    title: 'Veteran',
  },
};

export function grantAchievement(
  earned: string[],
  id: string,
): AchievementDef | null {
  if (earned.includes(id)) return null;
  earned.push(id);
  return ACHIEVEMENTS[id] ?? null;
}

File: crates/realm-core/src/achievements.rs
use std::collections::HashMap;
use std::sync::OnceLock;

#[derive(Debug, Clone)]
pub struct AchievementDef {
    pub id: &'static str,
    pub name: &'static str,
    pub description: &'static str,
    pub title: Option<&'static str>,
}

pub fn achievements() -> &'static HashMap<&'static str, AchievementDef> {
    static ACHIEVEMENTS: OnceLock<HashMap<&'static str, AchievementDef>> = OnceLock::new();
    ACHIEVEMENTS.get_or_init(|| {
        HashMap::from([
            (
                "first_blood",
                AchievementDef {
                    id: "first_blood",
                    name: "First Blood",
                    description: "Slay your first creature.",
                    title: Some("Initiate"),
                },
            ),
            (
                "goblin_hunter",
                AchievementDef {
                    id: "goblin_hunter",
                    name: "Goblin Hunter",
                    description: "Slay 10 goblins.",
                    title: Some("Goblin Hunter"),
                },
            ),
            (
                "chief_slayer",
                AchievementDef {
                    id: "chief_slayer",
                    name: "Kingslayer",
                    description: "Defeat Goblin Chief Grak.",
                    title: Some("Giantkiller"),
                },
            ),
            (
                "duelist",
                AchievementDef {
                    id: "duelist",
                    name: "Duelist",
                    description: "Win your first duel.",
                    title: Some("Duelist"),
                },
            ),
            (
                "explorer",
                AchievementDef {
                    id: "explorer",
                    name: "Explorer",
                    description: "Visit the Crystal Cavern.",
                    title: Some("Pathfinder"),
                },
            ),
            (
                "wealthy",
                AchievementDef {
                    id: "wealthy",
                    name: "Wealthy",
                    description: "Accumulate 500 gold.",
                    title: Some("Magnate"),
                },
            ),
            (
                "veteran",
                AchievementDef {
                    id: "veteran",
                    name: "Veteran",
                    description: "Reach level 10.",
                    title: Some("Veteran"),
                },
            ),
        ])
    })
}

pub fn grant_achievement(earned: &mut Vec<String>, id: &str) -> Option<AchievementDef> {
    if earned.iter().any(|e| e == id) {
        return None;
    }

    earned.push(id.to_string());
    achievements().get(id).cloned()
}